lib/logger/lib/logger/config.ex

```diff
@@ -1,95 +1,72 @@
-
   @behaviour :gen_event
   @name __MODULE__
-  @table __MODULE__
-  @data_key __MODULE__
-  @deleted_handlers_key __MODULE__.DeletedHandlers
-  @counter_pos 1
   @update_counter_message {__MODULE__, :update_counter}
 
+  alias Logger.Counter
+
   def configure(options) do
     :gen_event.call(Logger, @name, {:configure, options})
   end
 
+  def default_config do
+    sync_threshold = Application.fetch_env!(:logger, :sync_threshold)
+    discard_threshold = Application.fetch_env!(:logger, :discard_threshold)
+    level = Application.fetch_env!(:logger, :level)
+    sasl_reports? = Application.fetch_env!(:logger, :handle_sasl_reports)
+
+    %{
+      level: level,
+      utc_log: Application.fetch_env!(:logger, :utc_log),
+      truncate: Application.fetch_env!(:logger, :truncate),
+      translators: Application.fetch_env!(:logger, :translators),
+      thresholds: {sync_threshold, discard_threshold},
+      sasl: sasl_reports?
+    }
+  end
+
   def add_translator(translator) do
-    :gen_event.call(Logger, @name, {:add_translator, translator})
+    update_translators(fn t -> [translator | List.delete(t, translator)] end)
   end
 
   def remove_translator(translator) do
-    :gen_event.call(Logger, @name, {:remove_translator, translator})
+    update_translators(&List.delete(&1, translator))
   end
 
   def level do
-    %{level: level} = read_data!(@data_key)
+    %{level: level} = :logger.get_primary_config()
+
     level
   end
 
   def compare_levels(level, level), do: :eq
+  def compare_levels(:all, _), do: :gt
+  def compare_levels(_, :all), do: :lt
+  def compare_levels(:none, _), do: :lt
+  def compare_levels(_, :none), do: :gt
 
   def compare_levels(left, right) do
-    if level_to_number(left) > level_to_number(right), do: :gt, else: :lt
-  end
-
-  defp level_to_number(:debug), do: 0
-  defp level_to_number(:info), do: 1
-  defp level_to_number(:warn), do: 2
-  defp level_to_number(:error), do: 3
-
-  def translation_data do
-    read_data!(@data_key)
+    :logger.compare_levels(normalise(left), normalise(right))
   end
 
-  def log_data(level) do
-    %{level: min_level} = config = read_data!(@data_key)
+  defp normalise(:warn), do: :warning
+  defp normalise(level), do: level
 
-    if compare_levels(level, min_level) != :lt do
-      %{thresholds: {counter, sync, discard}} = config
-      value = bump_counter(counter)
-
-      cond do
-        value >= discard -> {:discard, config}
-        value >= sync -> {:sync, config}
-        true -> {:async, config}
-      end
-    else
-      {:discard, config}
-    end
-  end
-
-  def deleted_handlers do
-    try do
-      read_data(@deleted_handlers_key, [])
-    rescue
-      ArgumentError -> []
-    end
-  end
-
-  def deleted_handlers(handlers) do
-    :gen_event.call(Logger, @name, {:deleted_handlers, handlers})
-  end
-
-  def new do
-    {new_data(), new_counter()}
-  end
+  # TODO: Use counters exclusively when we require Erlang/OTP 22+.
+  defdelegate new, to: Logger.Counter
```

Comment:
Maybe we should remove this and call `Logger.Counter` directly in the appropriate places?

Reply:
We can, I did it to reduce amount of changes, but it shouldn't change much.

Reply:
@hauleth yeah, the counter is treated especially everywhere in this new implementation, so I think it makes sense.